File: source-code/Java-Spring/src/main/java/com/aadb2c/webapi/IdentityController.java
package com.aadb2c.webapi;

import java.util.Map;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class IdentityController {
    @GetMapping("/")
    public String welcome() {
        return "Welcome to Azure AD B2C custom REST API";
    }

    @PostMapping("/api/identity/loyalty")
    public ResponseEntity<B2CResponseModel> loyalty(@RequestBody(required = false) Map<String, String> body) {

        // Check if the POST data is provided
        if (body == null || body.isEmpty()) {
            return ResponseEntity.status(HttpStatus.CONFLICT).body(new B2CResponseError("Request content is null"));
        }

        // Check if the language parameter is provided
        if (body.containsKey("language") == false || body.get("language").isEmpty()) {
            return ResponseEntity.status(HttpStatus.CONFLICT)
                    .body(new B2CResponseError("Language code is null or empty"));
        }

        // Check if the objectId parameter is provided
        if (body.containsKey("objectId") == false || body.get("objectId").isEmpty()) {
            return ResponseEntity.status(HttpStatus.CONFLICT)
                    .body(new B2CResponseError("User object Id is null or empty"));
        }

        return ResponseEntity
                .ok(new B2CResponse(body.get("language") + "-" + ((int) (Math.random() * 9999 + 1000)), null));
    }

    @PostMapping("/api/identity/validate")
    public ResponseEntity<B2CResponseModel> validate(@RequestBody(required = false) Map<String, String> body) {

        // Check if the POST data is provided
        if (body == null || body.isEmpty()) {
            return ResponseEntity.status(HttpStatus.CONFLICT).body(new B2CResponseError("Request content is null"));
        }

        // Check if the language parameter is provided
        if (body.containsKey("language") == false || body.get("language").isEmpty()) {
            return ResponseEntity.status(HttpStatus.CONFLICT)
                    .body(new B2CResponseError("Language code is null or empty"));
        }

        // Check if the email parameter is provided
        if (body.containsKey("email") == false || body.get("email").isEmpty()) {
            return ResponseEntity.status(HttpStatus.CONFLICT).body(new B2CResponseError("Email is null or empty"));
        }

        // Check if the email parameter starts with 'test'
        if (body.get("email").toLowerCase().startsWith("test")) {
            return ResponseEntity.status(HttpStatus.CONFLICT)
                    .body(new B2CResponseError("Your email address cannot start with 'test'"));
        }

        return ResponseEntity.ok(new B2CResponse(body.get("language") + "-" + ((int) (Math.random() * 9999 + 1000)),
                body.get("email").toLowerCase()));
    }
}

File: source-code/dot-net-fw/AADB2C.WebAPI/Controllers/IdentityController.cs
using AADB2C.WebAPI.Models;
using Newtonsoft.Json;
using System;
using System.Net;
using System.Web.Http;

namespace Contoso.AADB2C.APIBasicAuth.Controllers
{
    public class IdentityController : ApiController
    {
        Random rnd = new Random();

        [Route("api/identity/loyalty")]
        [HttpPost]
        public IHttpActionResult Loyalty()
        {
            // If not data came in, then returen
            if (this.Request.Content == null) throw new Exception();

            // Read the input claims from the request body
            string input = Request.Content.ReadAsStringAsync().Result;

            // Check input content value
            if (string.IsNullOrEmpty(input))
            {
                return Content(HttpStatusCode.Conflict, new B2CResponseModel("Request content is empty", HttpStatusCode.Conflict));
            }

            // Convert the input string into InputClaimsModel object
            InputClaimsModel inputClaims = JsonConvert.DeserializeObject(input, typeof(InputClaimsModel)) as InputClaimsModel;

            if (inputClaims == null)
            {
                return Content(HttpStatusCode.Conflict, new B2CResponseModel("Can not deserialize input claims", HttpStatusCode.Conflict));
            }

            //Check if the language parameter is presented
            if (string.IsNullOrEmpty(inputClaims.language))
            {
                return Content(HttpStatusCode.Conflict, new B2CResponseModel("Language code is null or empty", HttpStatusCode.Conflict));
            }

            //Check if the objectId parameter is presented
            if (string.IsNullOrEmpty(inputClaims.objectId))
            {
                return Content(HttpStatusCode.Conflict, new B2CResponseModel("User object Id is null or empty", HttpStatusCode.Conflict));
            }

            // Create output claims object and set the loyalty number with random value
            B2CResponseModel outputClaims = new B2CResponseModel(string.Empty, HttpStatusCode.OK);
            outputClaims.loyaltyNumber = inputClaims.language + "-" + rnd.Next(1000, 9999).ToString();

            // Return the output claim(s)
            return Ok(outputClaims);
        }

        [Route("api/identity/Validate")]
        [HttpPost]
        public IHttpActionResult Validate()
        {
            // If not data came in, then returen
            if (this.Request.Content == null) throw new Exception();

            // Read the input claims from the request body
            string input = Request.Content.ReadAsStringAsync().Result;

            // Check input content value
            if (string.IsNullOrEmpty(input))
            {
                return Content(HttpStatusCode.Conflict, new B2CResponseModel("Request content is empty", HttpStatusCode.Conflict));
            }

            // Convert the input string into InputClaimsModel object
            InputClaimsModel inputClaims = JsonConvert.DeserializeObject(input, typeof(InputClaimsModel)) as InputClaimsModel;

            if (inputClaims == null)
            {
                return Content(HttpStatusCode.Conflict, new B2CResponseModel("Can not deserialize input claims", HttpStatusCode.Conflict));
            }

            //Check if the language parameter is presented
            if (string.IsNullOrEmpty(inputClaims.language))
            {
                return Content(HttpStatusCode.Conflict, new B2CResponseModel("Language code is null or empty", HttpStatusCode.Conflict));
            }

            //Check if the email parameter is presented
            if (string.IsNullOrEmpty(inputClaims.email))
            {
                return Content(HttpStatusCode.Conflict, new B2CResponseModel("Email is null or empty", HttpStatusCode.Conflict));
            }

            // Validate the email address 
            if (inputClaims.email.ToLower().StartsWith("test"))
            {
                return Content(HttpStatusCode.Conflict, new B2CResponseModel("Your email address can't start with 'test'", HttpStatusCode.Conflict));
            }

            // Create output claims object and set the loyalty number with random value
            B2CResponseModel outputClaims = new B2CResponseModel(string.Empty, HttpStatusCode.OK);
            outputClaims.loyaltyNumber = inputClaims.language + "-" + rnd.Next(1000, 9999).ToString();
            outputClaims.email = inputClaims.email.ToLower();

            // Return the output claim(s)
            return Ok(outputClaims);
        }
    }
}
